pandas/io/sql.py

```diff
@@ -1483,7 +1496,7 @@ def to_sql(
         if dtype and not is_dict_like(dtype):
             dtype = {col_name: dtype for col_name in frame}
 
-        if dtype is not None:
+        if dtype is not None and isinstance(dtype, dict):
```

Comment:
interesting you have to do this as L1496 explicity converts this.

Reply:
so is_dict_like will pass thrue a Series for example which will fail in other places which we are not likely testing. I would  change L1499 to
```
if dtype is not None:
     if not is_dict_like(...):
       ..
    else:
        dtype = dict(dtype)
```

Reply:
> interesting you have to do this as L1496 explicity converts this.

Yes, the problem is that we define the type of `dtype` already in the function at L1452. After that you cannot change the type of `dtype`, even not by overwriting the variable. `isinstance` checks are the [only way](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-attribute-exists-in-each-union-item-union-attr) to narrow it down. So the provided solution (see below) will not work, since we are overwriting the `dtype` variable

```python
        if dtype is not None:
            if not is_dict_like(dtype):
                dtype = {col_name: dtype for col_name in frame}
            else:
                dtype = dict(dtype) # This line gives a mypy error
```

Mypy error
```python
error: Argument 1 to "dict" has incompatible type "Union[ExtensionDtype, Any, str, Type[object], Dict[Optional[Hashable], Union[ExtensionDtype, str, Any, Type[str], Type[float], Type[int], Type[complex], Type[bool], Type[object]]]]"; expected "Mapping[Any, Any]"  [arg-type]
```

Reply:
> `isinstance` checks are the [only way](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-attribute-exists-in-each-union-item-union-attr) to narrow it down.

also can use cast or assert. In this case, is_dict_like function will not narrow types, so ok to use a cast following the is_* call. so something like

```
            if is_dict_like(dtype):
                dtype = cast(dict, dtype)
                ...
            else:
                ...
```

can always replace dict with Mapping, or union if more than dict is accepted for dict-like parameter.

Reply:
Thanks @simonjayhawkins did not know the `cast` function also helps in narrowing the type down. With the `cast` function it works.